File: Functions/extractGITTfromTrace.m
function gittEpisodes = extractGITTfromTrace(timeWithGaps, voltage, current, timeS, params)
% extractGITTfromTrace - find back-to-back discharge+charge GITT episodes
% in an ageing trace and extract the per-pulse OCV (voltage at the end of
% the rest that follows each pulse).
%
% Inputs
%   timeWithGaps - datetime array (may contain NaT at gap markers)
%   voltage      - V array (NaN at gaps)
%   current      - A array (NaN at gaps); sign convention: + charge / - discharge
%   timeS        - seconds-since-start array (NaN at gaps)
%   params       - struct with fields:
%       pulseAmp_A           - legacy / documentation only (nominal C/5 = 11.6 A)
%       pulseTol_A           - legacy / documentation only
%       maxPulseAmp_A        - UPPER |I| limit defining a pulse (A); the
%                              lower limit is restThr_A. Pulses with
%                              restThr_A < |I| < maxPulseAmp_A are kept
%                              (so C/5 and lower-rate GITT pulses are
%                              detected, but cycling at -87 / +29 A is
%                              excluded) (A-001)
%       minPulse_s           - reject pulses shorter than this (noise)
%       maxPulse_s           - reject pulses longer than this (cycling/checkup)
%       maxIntraEpisodeGap_s - max time gap between consecutive pulses inside
%                              one GITT episode (rests are <= this gap).
%                              Set to ~5 days so the discharge-GITT and
%                              charge-GITT halves of one checkup merge into
%                              a single episode while consecutive checkups
%                              (~3 weeks apart) stay separate (A-001).
%       restThr_A            - |I| below this is considered rest when
%                              locating the OCV sample after the last pulse (A-001)
%       minRestDur_s         - minimum contiguous rest duration accepted as
%                              the post-last-pulse OCV anchor (A-001)
%       pulseFlatnessTol     - max std(I)/|mean(I)| within a pulse; rejects
%                              CV-decay phases that ramp through the
%                              [restThr_A, maxPulseAmp_A] amplitude window
%                              during the initial CLUSTERING step (A-001)
%       gradThr_Apers        - |dI/dt| threshold (A/s) used inside a
%                              validated cluster to re-detect pulse rising
%                              edges. Once an episode is gated by
%                              clustering, refinement only needs to spot
%                              the sharp ramp at the start of each pulse;
%                              this avoids depending on pulse shape
%                              (flatness) so a CC+CV last pulse is still
%                              picked up (A-001)
%       minPulsesPerEpisode  - keep only clusters with at least this many pulses
%       maxPulsesPerEpisode  - reject clusters with more than this many
%                              pulses (a real GITT half has 23-26 C/5 pulses;
%                              a merged discharge+charge episode is ~46-52;
%                              this upper bound rejects cycling clusters
%                              that happen to slip past the flatness check) (A-001)
%
% Output
%   gittEpisodes - struct array; one element per detected episode:
%       .timeStart, .timeEnd      - episode time bounds (datetime)
%       .pulseStartTimes (Nx1)    - datetime
%       .pulseEndTimes   (Nx1)    - datetime
%       .pulseStartIdx,  (Nx1)    - sample index of each pulse start
%       .pulseEndIdx     (Nx1)    - sample index of each pulse end
%       .pulseModes      (Nx1 cell of 'discharge'/'charge')
%       .OCV_V          ((N+1)x1) - OCV at end of each rest (including pre-first)
%       .ocvIdx         ((N+1)x1) - sample index of each OCV measurement
%       .ocvModes       ((N+1)x1) - 'discharge' / 'charge' half-cycle of each OCV
%       .cumQ_signed_Ah ((N+1)x1) - cumulative signed charge since episode start
%                                   (starts at 0 BEFORE the first pulse) (A-001)
%       .q_per_pulse_Ah  (Nx1)    - signed Ah delivered by each pulse
%
% Author: GitHub Copilot (for Feye Hoekstra)
% Date:   2026-07-28 (extracted from PlotCellSummary.m into a shared helper)
% Assumption: A-001 (see docs/assumptions.md)

% Defensive defaults (in case caller passes a partial struct)
def = struct('pulseAmp_A',11.6, 'pulseTol_A',0.5, ...
             'maxPulseAmp_A',20, ...
             'minPulse_s',60, 'maxPulse_s',3600, ...
             'maxIntraEpisodeGap_s',24*3600, 'restThr_A',0.5, ...
             'minRestDur_s',600, 'pulseFlatnessTol',0.05, ...
             'gradThr_Apers',3, ...
             'minPulsesPerEpisode',20, 'maxPulsesPerEpisode',55);
fn = fieldnames(def);
for k = 1:numel(fn)
    if ~isfield(params, fn{k}) || isempty(params.(fn{k}))
        params.(fn{k}) = def.(fn{k});
    end
end

% Empty default output (compatible with [] short-circuit downstream)
gittEpisodes = struct([]);

% 1) Pulse mask: any current excursion above rest and below the cycling
%    amplitude. This captures BOTH the C/5 GITT pulses and any lower-rate
%    (C/10, C/20) pulses used at the discharge cutoff (A-001). The duration
%    filter below removes continuous C/5 references and CV-charge phases.
pulseMask = (abs(current) > params.restThr_A) ...
          & (abs(current) < params.maxPulseAmp_A) ...
          & ~isnan(current);

% 2) Find continuous pulse segments (runs of true)
%    Use a 0/1 diff trick; works for both row and column vectors.
pm = pulseMask(:);
edges = diff([false; pm; false]);
pulseStarts = find(edges ==  1);
pulseEnds   = find(edges == -1) - 1;

if isempty(pulseStarts); return; end

% 3) Filter pulses by duration and build a pulse table
pulseStartIdx = [];
pulseEndIdx   = [];
pulseStartTime= NaT(0);
pulseEndTime  = NaT(0);
pulseModeCell = {};
for i = 1:numel(pulseStarts)
    s = pulseStarts(i); e = pulseEnds(i);
    if isnan(timeS(s)) || isnan(timeS(e)); continue; end
    dur = timeS(e) - timeS(s);
    if dur < params.minPulse_s || dur > params.maxPulse_s; continue; end
    % Constant-current flatness check: reject CV-decay phases of cycling
    % charge (current ramps through the [restThr_A, maxPulseAmp_A] window).
    iPulse = current(s:e);
    iPulse = iPulse(~isnan(iPulse));
    if isempty(iPulse); continue; end
    mu = abs(mean(iPulse));
    if mu < params.restThr_A; continue; end
    if (std(iPulse) / mu) > params.pulseFlatnessTol; continue; end
    if mean(current(s:e), 'omitnan') < 0
        m = 'discharge';
    else
        m = 'charge';
    end
    pulseStartIdx(end+1,1)  = s; %#ok<AGROW>
    pulseEndIdx(end+1,1)    = e; %#ok<AGROW>
    pulseStartTime(end+1,1) = timeWithGaps(s); %#ok<AGROW>
    pulseEndTime(end+1,1)   = timeWithGaps(e); %#ok<AGROW>
    pulseModeCell{end+1,1}  = m; %#ok<AGROW>
end
nPulses = numel(pulseStartIdx);
if nPulses == 0; return; end

% 4) Cluster pulses into episodes by inter-pulse time gap
clusterId = ones(nPulses, 1);
for i = 2:nPulses
    gap_s = seconds(pulseStartTime(i) - pulseEndTime(i-1));
    if gap_s > params.maxIntraEpisodeGap_s
        clusterId(i:end) = clusterId(i:end) + 1;
    end
end

% 5) Build episode list for clusters with pulse counts in the expected range
nClusters = max(clusterId);
for c = 1:nClusters
    idx = find(clusterId == c);
    if numel(idx) < params.minPulsesPerEpisode; continue; end
    if numel(idx) > params.maxPulsesPerEpisode; continue; end

    % Per-half pulse-count sanity check. A real GITT half (discharge OR
    % charge) has 23-26 C/5 pulses. A merged episode contains two such
    % halves separated by a multi-hour rest (~10 h on this cell). Detect
    % half boundaries inside the cluster as the largest inter-pulse rest(s)
    % and require every half to fall in [minHalf, maxHalf]. This rejects
    % clusters whose total count happens to land in [min, max] but whose
    % per-half breakdown is wrong (e.g., a few spurious pulses + a real
    % half, or a half-checkup spliced to an unrelated cycling pulse). (A-001)
    minHalf = 23 - 3;   % allow 3 missing/rejected pulses per half
    maxHalf = 26 + 3;   % allow 3 spurious accepted pulses per half
    intraGaps_s = seconds(pulseStartTime(idx(2:end)) - pulseEndTime(idx(1:end-1)));
    % A rest is a "half boundary" if it is much longer than a normal
    % within-half rest (~2 h). Use 4 h as a robust threshold: comfortably
    % above the typical 2 h GITT rest and below the ~10 h between-half gap.
    halfBoundaryThr_s = 4 * 3600;
    halfBoundaries = find(intraGaps_s > halfBoundaryThr_s);
    halfStartsLocal = [1; halfBoundaries + 1];
    halfEndsLocal   = [halfBoundaries; numel(idx)];
    halfSizes = halfEndsLocal - halfStartsLocal + 1;
    if any(halfSizes < minHalf) || any(halfSizes > maxHalf)
        fprintf('  [GITT] Rejected cluster (%s -> %s, %d pulses, halves = [%s]).\n', ...
            datestr(pulseStartTime(idx(1)),   'yyyy-mm-dd'), ...
            datestr(pulseEndTime(idx(end)),   'yyyy-mm-dd'), ...
            numel(idx), strjoin(string(halfSizes), ','));
        continue
    end

    % --- Refinement inside the validated cluster --------------------
    % The cluster step uses a CC flatness check, which (correctly) rejects
    % cycling CV-decay phases but ALSO rejects the last GITT pulse of a
    % discharge/charge half (CC + CV taper). And the post-last-pulse OCV
    % anchor used a "first rest >= minRestDur_s" walk-forward, which can
    % latch onto an intermediate rest rather than the rest after the true
    % final pulse. Both problems disappear with a simpler approach now
    % that we KNOW the cluster is a real GITT episode (A-001):
    %   - Within the cluster's time window (extended forward to capture
    %     a CV-tapered last pulse + its long rest), find every pulse
    %     RISING edge by |dI/dt| > gradThr_Apers. A C/5 step ramps at
    %     ~11.6 A/s; CV taper and noise stay well below the threshold.
    %   - Each rest OCV is the sample just before a rising edge. The
    %     post-last-pulse OCV is the sample just before the FIRST rising
    %     edge that occurs AFTER the refined last pulse ends (i.e., the
    %     start of the next test step / next checkup activity).
    %   - Each pulse end is found by walking forward from the rising
    %     edge until |I| drops below restThr_A (handles CC + optional CV).
    sFirst = pulseStartIdx(idx(1));
    sLastClusterPulse = pulseEndIdx(idx(end));
    % Walk the window start backwards into the pre-pulse rest so that
    % iSeg(1) is a RESTING sample. Without this, pulse 1 starts at index
    % 1 of seg, restMask(1) is false (the cell is already pulsing), and
    % the rising-edge detector misses the first pulse entirely. Walk back
    % through any contiguous rest until we hit a non-rest sample or a
    % NaN gap; cap at index 1.
    while sFirst > 1
        if isnan(current(sFirst - 1)) || abs(current(sFirst - 1)) >= params.restThr_A
            break
        end
        sFirst = sFirst - 1;
    end
    % Extend lookahead by one inter-episode gap so a CV-tapered last
    % pulse and the long rest that follows are inside the window.
    tEndSearch  = timeS(sLastClusterPulse) + params.maxIntraEpisodeGap_s;
    candidate = find(~isnan(timeS(:)) & timeS(:) <= tEndSearch);
    if isempty(candidate)
        sLast = numel(timeS);
    else
        sLast = max(candidate);
    end
    if sLast < sFirst; sLast = numel(timeS); end

    seg  = (sFirst:sLast).';
    iSeg = current(seg);
    tSeg = timeS(seg);
    dI = diff(iSeg);
    dT = diff(tSeg);
    diDt = dI ./ dT;            % signed A/s (NaN where iSeg/tSeg has NaN)
    absDiDt = abs(diDt);

    % A rising edge is a sample where current changes sharply AND the
    % cell was at rest just before (so we don't pick up the down-ramp at
    % the end of a pulse, or mid-pulse noise).
    restMask = abs(iSeg(1:end-1)) < params.restThr_A;
    isEdge   = (absDiDt > params.gradThr_Apers) & restMask;
    isEdge(isnan(isEdge)) = false;

    % Each ramp produces one run of true in isEdge (because once the
    % current is high, restMask flips false). Take the START of each run.
    edgeStarts = find(diff([false; isEdge(:)]) == 1);
    if isempty(edgeStarts); continue; end

    prePulseGlobal  = seg(edgeStarts);       % last resting sample (= OCV anchor)
    pulseStartIdx_r = seg(edgeStarts + 1);   % first ramp sample (= pulse start)

    % Per-pulse mode from the sign of dI/dt at the edge.
    pulseModes_r = cell(numel(edgeStarts), 1);
    for k = 1:numel(edgeStarts)
        if diDt(edgeStarts(k)) < 0
            pulseModes_r{k} = 'discharge';
        else
            pulseModes_r{k} = 'charge';
        end
    end

    % Pulse end: walk forward from pulseStartIdx_r(k) until |I| returns
    % below restThr_A or until the next pulse starts. Handles CC + CV.
    pulseEndIdx_r = zeros(numel(edgeStarts), 1);
    for k = 1:numel(edgeStarts)
        if k < numel(edgeStarts)
            walkEnd = prePulseGlobal(k + 1);
        else
            walkEnd = sLast;
        end
        pe = pulseStartIdx_r(k);
        for j = pulseStartIdx_r(k):walkEnd
            if isnan(current(j)) || abs(current(j)) < params.restThr_A
                pe = j - 1;
                break
            end
            pe = j;
        end
        pulseEndIdx_r(k) = pe;
    end

    % Truncate to the episode boundary: stop once the inter-pulse gap
    % exceeds maxIntraEpisodeGap_s (we walked into the next episode).
    nKeep = numel(pulseStartIdx_r);
    for k = 2:numel(pulseStartIdx_r)
        gap_s = timeS(pulseStartIdx_r(k)) - timeS(pulseEndIdx_r(k - 1));
        if ~isnan(gap_s) && gap_s > params.maxIntraEpisodeGap_s
            nKeep = k - 1;
            break
        end
    end
    pulseStartIdx_r = pulseStartIdx_r(1:nKeep);
    pulseEndIdx_r   = pulseEndIdx_r(1:nKeep);
    pulseModes_r    = pulseModes_r(1:nKeep);
    prePulseGlobal  = prePulseGlobal(1:nKeep);

    nP = numel(pulseStartIdx_r);
    if nP < params.minPulsesPerEpisode || nP > params.maxPulsesPerEpisode
        fprintf('  [GITT] Refinement rejected cluster (%s, %d pulses out of [%d,%d]).\n', ...
            datestr(timeWithGaps(sFirst), 'yyyy-mm-dd'), nP, ...
            params.minPulsesPerEpisode, params.maxPulsesPerEpisode);
        continue
    end

    % Per-pulse signed Ah (trapezoidal over [pulseStart, pulseEnd]).
    q_per_pulse_Ah = zeros(nP, 1);
    for k = 1:nP
        s = pulseStartIdx_r(k);
        e = pulseEndIdx_r(k);
        tk = timeS(s:e); ik = current(s:e);
        m = ~isnan(tk) & ~isnan(ik);
        if nnz(m) >= 2
            q_per_pulse_Ah(k) = trapz(tk(m), ik(m)) / 3600;
        end
    end

    % OCV anchors: prePulseGlobal(k) for k=1..nP, plus the rest after the
    % last pulse, anchored at the FIRST rising edge AFTER pulseEndIdx_r(nP).
    ocvIdx    = zeros(nP + 1, 1);
    ocvModesC = cell(nP + 1, 1);
    ocvIdx(1)    = prePulseGlobal(1);
    ocvModesC{1} = pulseModes_r{1};
    for k = 1:nP - 1
        ocvIdx(k + 1)    = prePulseGlobal(k + 1);
        ocvModesC{k + 1} = pulseModes_r{k};
    end
    % Post-last-pulse anchor: scan from pulseEndIdx_r(nP)+1 to end of trace.
    lastE = pulseEndIdx_r(nP);
    ocvIdx(nP + 1)    = lastE;            % fallback (trace cut short)
    ocvModesC{nP + 1} = pulseModes_r{nP};
    rem = (lastE + 1 : numel(timeS)).';
    if numel(rem) >= 2
        iR = current(rem);
        tR = timeS(rem);
        diDtR = diff(iR) ./ diff(tR);
        restMaskR = abs(iR(1:end-1)) < params.restThr_A;
        isEdgeR = (abs(diDtR) > params.gradThr_Apers) & restMaskR;
        isEdgeR(isnan(isEdgeR)) = false;
        firstEdgeR = find(isEdgeR, 1, 'first');
        if ~isempty(firstEdgeR)
            ocvIdx(nP + 1) = rem(firstEdgeR);
        end
    end

    OCV_V = voltage(ocvIdx);

    ep = struct();
    ep.timeStart       = timeWithGaps(pulseStartIdx_r(1));
    ep.timeEnd         = timeWithGaps(pulseEndIdx_r(end));
    ep.pulseStartTimes = timeWithGaps(pulseStartIdx_r);
    ep.pulseEndTimes   = timeWithGaps(pulseEndIdx_r);
    ep.pulseStartIdx   = pulseStartIdx_r;
    ep.pulseEndIdx     = pulseEndIdx_r;
    ep.pulseModes      = pulseModes_r;
    ep.cumQ_signed_Ah  = [0; cumsum(q_per_pulse_Ah)];
    ep.OCV_V           = OCV_V;
    ep.ocvIdx          = ocvIdx;
    ep.ocvModes        = ocvModesC;
    ep.q_per_pulse_Ah  = q_per_pulse_Ah;

    if isempty(fieldnames(gittEpisodes))
        gittEpisodes = ep;
    else
        gittEpisodes(end+1) = ep; %#ok<AGROW>
    end
end
end

File: Functions/extractGITTfromTrace.py
"""
extractGITTfromTrace - Find back-to-back discharge+charge GITT episodes in
an ageing trace and extract the per-pulse OCV.

Python counterpart of Functions/extractGITTfromTrace.m, used by
python_scripts/PlotCellSummary.py (todo #012).

Author: GitHub Copilot (for Robinson Medina / Feye Hoekstra)
Date: 2026-08-24
Assumption: A-001 (see docs/assumptions.md)

Compliance: R-004, R-012.
"""

import numpy as np

try:
    # NumPy >= 2.0 name (np.trapz was deprecated in 2.0 and removed in 2.5).
    _trapz = np.trapezoid
except AttributeError:  # pragma: no cover - NumPy < 2.0 fallback
    _trapz = np.trapz

_DEFAULTS = {
    'pulseAmp_A': 11.6, 'pulseTol_A': 0.5, 'maxPulseAmp_A': 20,
    'minPulse_s': 60, 'maxPulse_s': 3600, 'minPulseCharge_Ah': 0.15,
    'maxIntraEpisodeGap_s': 24 * 3600, 'restThr_A': 0.5,
    'minRestDur_s': 600, 'pulseFlatnessTol': 0.05, 'edgeJump_A': 2,
    'gradThr_Apers': 3, 'minPulsesPerEpisode': 20, 'maxPulsesPerEpisode': 55,
}


def _find_runs(mask):
    """Return (starts, ends) 0-based inclusive index arrays of contiguous True runs."""

    padded = np.concatenate(([False], mask, [False])).astype(int)
    dd = np.diff(padded)
    starts = np.where(dd == 1)[0]
    ends = np.where(dd == -1)[0] - 1
    return starts, ends


def extract_gitt_from_trace(time_with_gaps, voltage, current, time_s, params):
    """
    Detect GITT episodes (clusters of C/5 +/- pulses) and extract per-pulse OCV.

    Returns a list of dicts, one per detected episode, with keys:
    time_start, time_end, pulse_start_times, pulse_end_times,
    pulse_start_idx, pulse_end_idx, pulse_modes, cumQ_signed_Ah, OCV_V,
    ocv_idx, ocv_modes, q_per_pulse_Ah.
    """

    p = dict(_DEFAULTS)
    p.update({k: v for k, v in params.items() if v is not None})

    gitt_episodes = []
    time_s = np.asarray(time_s, dtype=float)
    current = np.asarray(current, dtype=float)
    voltage = np.asarray(voltage, dtype=float)
    n = len(current)

    # 1) Pulse mask: current excursion above rest and below cycling amplitude.
    pulse_mask = (np.abs(current) > p['restThr_A']) & (np.abs(current) < p['maxPulseAmp_A']) & ~np.isnan(current)
    pulse_starts, pulse_ends = _find_runs(pulse_mask)
    if len(pulse_starts) == 0:
        return gitt_episodes

    # 3) Filter pulses by duration + peak amplitude + charge; build the pulse table.
    pulse_start_idx, pulse_end_idx, pulse_modes = [], [], []
    for s, e in zip(pulse_starts, pulse_ends):
        if np.isnan(time_s[s]) or np.isnan(time_s[e]):
            continue
        dur = time_s[e] - time_s[s]
        if dur < p['minPulse_s'] or dur > p['maxPulse_s']:
            continue
        # Pulse-acceptance test (#082 fix, ported 2026-08-26 from
        # matlab_scripts/PlotCellSummary.m): a genuine C/5 GITT pulse can end
        # in a CC->CV taper near the SoC extremes (voltage cutoff hit
        # mid-pulse), which the old whole-pulse flatness test wrongly rejected
        # (confirmed on A3.10_Cell_22 EoL). Replaced with two taper-invariant
        # checks: (1) peak |I| must sit at the nominal C/5 amplitude - a CV
        # tail can only decay FROM that peak - and (2) the pulse must transfer
        # a minimum charge (real SoC change, not brief noise).
        i_pulse = current[s:e + 1]
        t_pulse = time_s[s:e + 1]
        keep_m = ~np.isnan(i_pulse) & ~np.isnan(t_pulse)
        i_pulse = i_pulse[keep_m]
        t_pulse = t_pulse[keep_m]
        if len(i_pulse) == 0:
            continue
        i_peak = np.max(np.abs(i_pulse))
        if abs(i_peak - p['pulseAmp_A']) > p['pulseTol_A']:
            continue
        if len(i_pulse) < 2:
            continue
        q_pulse_ah = _trapz(i_pulse, t_pulse) / 3600
        if abs(q_pulse_ah) < p['minPulseCharge_Ah']:
            continue
        mode = 'discharge' if np.nanmean(current[s:e + 1]) < 0 else 'charge'
        pulse_start_idx.append(s)
        pulse_end_idx.append(e)
        pulse_modes.append(mode)

    n_pulses = len(pulse_start_idx)
    if n_pulses == 0:
        return gitt_episodes

    pulse_start_idx = np.array(pulse_start_idx)
    pulse_end_idx = np.array(pulse_end_idx)
    pulse_start_time = time_with_gaps[pulse_start_idx]
    pulse_end_time = time_with_gaps[pulse_end_idx]

    # 4) Cluster pulses into episodes by inter-pulse time gap.
    cluster_id = np.zeros(n_pulses, dtype=int)
    for i in range(1, n_pulses):
        gap_s = (pulse_start_time[i] - pulse_end_time[i - 1]) / np.timedelta64(1, 's')
        cluster_id[i] = cluster_id[i - 1] + (1 if gap_s > p['maxIntraEpisodeGap_s'] else 0)

    # 5) Build episode list for clusters with pulse counts in range.
    min_half, max_half = 23 - 3, 26 + 3
    half_boundary_thr_s = 4 * 3600

    for c in range(cluster_id.max() + 1):
        idx = np.where(cluster_id == c)[0]
        if len(idx) < p['minPulsesPerEpisode'] or len(idx) > p['maxPulsesPerEpisode']:
            continue

        intra_gaps_s = (pulse_start_time[idx[1:]] - pulse_end_time[idx[:-1]]) / np.timedelta64(1, 's')
        half_boundaries = np.where(intra_gaps_s > half_boundary_thr_s)[0]
        half_starts_local = np.concatenate(([0], half_boundaries + 1))
        half_ends_local = np.concatenate((half_boundaries, [len(idx) - 1]))
        half_sizes = half_ends_local - half_starts_local + 1
        if np.any(half_sizes < min_half) or np.any(half_sizes > max_half):
            print(f"  [GITT] Rejected cluster ({pulse_start_time[idx[0]]} -> {pulse_end_time[idx[-1]]}, "
                  f"{len(idx)} pulses, halves = [{','.join(str(x) for x in half_sizes)}]).")
            continue

        # --- Refinement inside the validated cluster ------------------------
        s_first = pulse_start_idx[idx[0]]
        s_last_cluster_pulse = pulse_end_idx[idx[-1]]
        while s_first > 0:
            if np.isnan(current[s_first - 1]) or abs(current[s_first - 1]) >= p['restThr_A']:
                break
            s_first -= 1

        t_end_search = time_s[s_last_cluster_pulse] + p['maxIntraEpisodeGap_s']
        candidate = np.where(~np.isnan(time_s) & (time_s <= t_end_search))[0]
        s_last = candidate[-1] if len(candidate) else n - 1
        if s_last < s_first:
            s_last = n - 1

        seg = np.arange(s_first, s_last + 1)
        i_seg = current[seg]
        t_seg = time_s[seg]
        di_dt = np.diff(i_seg) / np.diff(t_seg)
        abs_di_dt = np.abs(di_dt)

        rest_mask = np.abs(i_seg[:-1]) < p['restThr_A']
        is_edge = (abs_di_dt > p['gradThr_Apers']) & rest_mask
        is_edge = np.nan_to_num(is_edge, nan=0.0).astype(bool)

        edge_run_starts, _ = _find_runs(is_edge)
        if len(edge_run_starts) == 0:
            continue

        pre_pulse_global = seg[edge_run_starts]           # last resting sample (OCV anchor)
        pulse_start_idx_r = seg[edge_run_starts + 1]       # first ramp sample (pulse start)

        pulse_modes_r = ['discharge' if di_dt[k] < 0 else 'charge' for k in edge_run_starts]

        # Pulse end: walk forward until |I| < restThr_A or the next pulse starts.
        pulse_end_idx_r = np.zeros(len(edge_run_starts), dtype=int)
        for k in range(len(edge_run_starts)):
            walk_end = pre_pulse_global[k + 1] if k < len(edge_run_starts) - 1 else s_last
            pe = pulse_start_idx_r[k]
            for j in range(pulse_start_idx_r[k], walk_end + 1):
                if np.isnan(current[j]) or abs(current[j]) < p['restThr_A']:
                    pe = j - 1
                    break
                pe = j
            pulse_end_idx_r[k] = pe

        # Truncate to the episode boundary: stop once the inter-pulse gap
        # exceeds maxIntraEpisodeGap_s (we walked into the next episode), OR
        # once a pulse's own duration/charge is far outside what a genuine
        # C/5 GITT pulse can be (found 2026-08-25, Cell_34, ported from the
        # MATLAB fix in matlab_scripts/PlotCellSummary.m): the refined
        # rising-edge scan above has NO amplitude/charge gate (unlike the
        # coarse pass), so within the maxIntraEpisodeGap_s lookahead window
        # it can pick up one extra unrelated pulse - e.g. the start of the
        # next test phase - as a spurious "one more GITT pulse". A nominal
        # full C/5 pulse carries ~2.42 Ah over ~12.5 min; allow a generous
        # margin so legitimately longer/CV-tapered pulses near the SoC
        # extremes still pass, while the much larger/longer next-phase pulse
        # does not. This MAXIMUM is a FIXED value, decoupled from the
        # per-pulse MINIMUM minPulseCharge_Ah (0.15 Ah, #082 2026-08-26).
        max_single_pulse_charge_ah = 3.63  # ~1.5x the nominal full C/5 pulse charge (2.42 Ah)
        max_single_pulse_dur_s = 30 * 60                         # 30 min
        n_keep = len(pulse_start_idx_r)
        for k in range(len(pulse_start_idx_r)):
            if k > 0:
                gap_s = time_s[pulse_start_idx_r[k]] - time_s[pulse_end_idx_r[k - 1]]
                if not np.isnan(gap_s) and gap_s > p['maxIntraEpisodeGap_s']:
                    n_keep = k
                    break
            s, e = pulse_start_idx_r[k], pulse_end_idx_r[k]
            dur_s = time_s[e] - time_s[s]
            tk, ik = time_s[s:e + 1], current[s:e + 1]
            m = ~np.isnan(tk) & ~np.isnan(ik)
            qk = _trapz(ik[m], tk[m]) / 3600 if np.count_nonzero(m) >= 2 else 0.0
            if (not np.isnan(dur_s) and dur_s > max_single_pulse_dur_s) or abs(qk) > max_single_pulse_charge_ah:
                n_keep = k
                break
        pulse_start_idx_r = pulse_start_idx_r[:n_keep]
        pulse_end_idx_r = pulse_end_idx_r[:n_keep]
        pulse_modes_r = pulse_modes_r[:n_keep]
        pre_pulse_global = pre_pulse_global[:n_keep]

        n_p = len(pulse_start_idx_r)
        if n_p < p['minPulsesPerEpisode'] or n_p > p['maxPulsesPerEpisode']:
            print(f"  [GITT] Refinement rejected cluster ({time_with_gaps[s_first]}, {n_p} pulses "
                  f"out of [{p['minPulsesPerEpisode']},{p['maxPulsesPerEpisode']}]).")
            continue

        # Per-pulse signed Ah (trapezoidal over [pulseStart, pulseEnd]).
        q_per_pulse_ah = np.zeros(n_p)
        for k in range(n_p):
            s, e = pulse_start_idx_r[k], pulse_end_idx_r[k]
            tk, ik = time_s[s:e + 1], current[s:e + 1]
            m = ~np.isnan(tk) & ~np.isnan(ik)
            if np.count_nonzero(m) >= 2:
                q_per_pulse_ah[k] = _trapz(ik[m], tk[m]) / 3600

        # OCV anchors.
        ocv_idx = np.zeros(n_p + 1, dtype=int)
        ocv_modes = [None] * (n_p + 1)
        ocv_idx[0] = pre_pulse_global[0]
        ocv_modes[0] = pulse_modes_r[0]
        for k in range(0, n_p - 1):
            ocv_idx[k + 1] = pre_pulse_global[k + 1]
            ocv_modes[k + 1] = pulse_modes_r[k]
        last_e = pulse_end_idx_r[-1]
        ocv_idx[n_p] = last_e  # fallback
        ocv_modes[n_p] = pulse_modes_r[-1]
        rem = np.arange(last_e + 1, n)
        if len(rem) >= 2:
            i_r = current[rem]
            t_r = time_s[rem]
            di_dt_r = np.diff(i_r) / np.diff(t_r)
            rest_mask_r = np.abs(i_r[:-1]) < p['restThr_A']
            is_edge_r = (np.abs(di_dt_r) > p['gradThr_Apers']) & rest_mask_r
            is_edge_r = np.nan_to_num(is_edge_r, nan=0.0).astype(bool)
            first_edge_r = np.where(is_edge_r)[0]
            if len(first_edge_r):
                ocv_idx[n_p] = rem[first_edge_r[0]]

        ocv_v = voltage[ocv_idx]
        ep = {
            'timeStart': time_with_gaps[pulse_start_idx_r[0]],
            'timeEnd': time_with_gaps[pulse_end_idx_r[-1]],
            'pulseStartTimes': time_with_gaps[pulse_start_idx_r],
            'pulseEndTimes': time_with_gaps[pulse_end_idx_r],
            'pulseStartIdx': pulse_start_idx_r,
            'pulseEndIdx': pulse_end_idx_r,
            'pulseModes': pulse_modes_r,
            'cumQ_signed_Ah': np.concatenate(([0.0], np.cumsum(q_per_pulse_ah))),
            'OCV_V': ocv_v,
            'ocvIdx': ocv_idx,
            'ocvModes': ocv_modes,
            'q_per_pulse_Ah': q_per_pulse_ah,
        }
        gitt_episodes.append(ep)

    return gitt_episodes


__all__ = ['extract_gitt_from_trace']
